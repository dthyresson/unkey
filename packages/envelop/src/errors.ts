export const NETWORK_ERROR = "Network error!";
export const RATE_LIMIT_ERROR = "Rate limit exceeded!";

export class NetworkError extends Error {
  constructor(message?: string) {
    super([NETWORK_ERROR, message].filter(Boolean).join(": "));
  }
}
export class RateLimitError extends Error {
  constructor(message?: string) {
    super([RATE_LIMIT_ERROR, message].filter(Boolean).join(": "));
  }
}
