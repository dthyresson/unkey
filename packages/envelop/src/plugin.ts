import { Plugin } from "@envelop/core";
import { type UnkeyError, verifyKey } from "@unkey/api";

export type UnkeyPluginOptions = { token: string; logger?: any };

/**
 * export type ErrorCode =
  | "NOT_FOUND"
  | "BAD_REQUEST"
  | "UNAUTHORIZED"
  | "INTERNAL_SERVER_ERROR"
  | "RATELIMITED"
  | "FORBIDDEN"
  | "KEY_USAGE_EXCEEDED"
  | "INVALID_KEY_TYPE"
  | "NOT_UNIQUE"
  | "FETCH_ERROR"; // not from unkey but returned when fetch fails

 */
export class NetworkError extends Error {}
export class RateLimitError extends Error {}

export const useUnkey = <TOptions extends UnkeyPluginOptions>(
  options: TOptions
): Plugin => {
  return {
    async onContextBuilding({
      context,
      extendContext,
    }: {
      context: Record<string, any>;
      extendContext: (context: Record<string, any>) => void;
    }) {
      try {
        const { result, error } = await verifyKey(options.token);

        if (error) {
          // handle potential network or bad request error
          // a link to our docs will be in the `error.docs` field
          console.error(error.message, error.code, error.requestId);
          throw new NetworkError(`Network error!`);
        }

        if (!result.valid) {
          throw new RateLimitError(`Rate limit exceeded!`);
        }
        extendContext({
          ["_unkey"]: result,
        });
      } catch (e) {
        throw e;
      }
    },
  };
};
