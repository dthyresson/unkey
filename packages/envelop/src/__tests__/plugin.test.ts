import { assertSingleExecutionValue, createTestkit } from "@envelop/testing";
import { makeExecutableSchema } from "@graphql-tools/schema";
import "dotenv/config";

import { NetworkError, RateLimitError } from "../errors";
import { useUnkey } from "../plugin";

describe("useUnkey", () => {
  const schema = makeExecutableSchema({
    typeDefs: /* GraphQL */ `
      type Query {
        foo: String!
      }
    `,
    resolvers: {
      Query: {
        foo: () => "hi",
      },
    },
  });

  describe("onContextBuilding", () => {
    it("Should query when not using the Unkey plugin", async () => {
      // Create a testkit for the plugin, using the plugin and a dummy schema
      const testkit = createTestkit([], schema);
      // Execute the envelop using a simple query
      const result = await testkit.execute("query { foo }");
      // During tests, it's simpler to assume you are dealing with a non-stream responses
      assertSingleExecutionValue(result);
      // Assert that the result is correct
      expect(result.errors).toBeUndefined();
      expect(result.data?.foo).toBe("hi");
    });

    it("Should error with invalid Unkey token", async () => {
      const testkit = createTestkit([useUnkey({ token: "123" })], schema);
      try {
        await testkit.execute("query { foo }");
      } catch (e) {
        expect(e).toBeInstanceOf(RateLimitError);
      }
    });

    it("Should error with invalid Unkey token", async () => {
      const testkit = createTestkit([useUnkey({ token: "" })], schema);
      try {
        await testkit.execute("query { foo }");
      } catch (e) {
        expect(e).toBeInstanceOf(NetworkError);
      }
    });

    it("Should rate limit", async () => {
      const token = process.env.RATE_LIMIT_5_PER_1_SECOND || "";
      const testkit = createTestkit([useUnkey({ token })], schema);
      try {
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        await testkit.execute("query { foo }");
        const result = await testkit.execute("query { foo }");
        assertSingleExecutionValue(result);
        // Assert that the result is correct
        expect(result.data).toBeUndefined();
      } catch (e) {
        expect(e).toBeInstanceOf(RateLimitError);
      }
    });
  });

  describe("errors", () => {
    describe("network errors", () => {
      it("A network error should have a default message", () => {
        const error = new NetworkError();
        expect(error.message).toMatchSnapshot();
      });
      it("A network error should have a custom message", () => {
        const error = new NetworkError("Custom message");
        expect(error.message).toMatchSnapshot();
      });
    });

    describe("rate limit errors", () => {
      it("A rate limit error should have a default message", () => {
        const error = new RateLimitError();
        expect(error.message).toMatchSnapshot();
      });
      it("A rate limit error should have a custom message", () => {
        const error = new RateLimitError("Custom message");
        expect(error.message).toMatchSnapshot();
      });
    });
  });
});
