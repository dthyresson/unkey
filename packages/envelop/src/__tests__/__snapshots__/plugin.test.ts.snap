// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useUnkey errors network errors A network error should have a custom message 1`] = `"Network error!: Custom message"`;

exports[`useUnkey errors network errors A network error should have a default message 1`] = `"Network error!"`;

exports[`useUnkey errors rate limit errors A rate limit error should have a custom message 1`] = `"Rate limit exceeded!: Custom message"`;

exports[`useUnkey errors rate limit errors A rate limit error should have a default message 1`] = `"Rate limit exceeded!"`;
